compiler/rustc_infer/src/infer/outlives/obligations.rs

```diff
@@ -392,23 +429,11 @@ where
         // the problem is to add `T: 'r`, which isn't true. So, if there are no
         // inference variables, we use a verify constraint instead of adding
         // edges, which winds up enforcing the same condition.
-        let needs_infer = projection_ty.needs_infer();
+        let needs_infer = substs.needs_infer();
         if approx_env_bounds.is_empty() && trait_bounds.is_empty() && needs_infer {
```

Comment:
The rationale for `&& need_infer` in the previous comment doesn't seem to apply to RPITs, because the type cannot be named anyway, so the user will  eventually need to specify the bounds on the components rater than on the opaque type.

This indeed causes diagnostics regression in this case:
```rust
fn id<T>(_: T) -> impl Sized {}
fn test<T>(t: T) -> impl Sized + 'static {
    id(t)
    //^ ERROR the opaque type `id<T>::{opaque#0}` may not live long enough
}
```

But removing `needs_infer` condition for RPITs may not be a good solution either since it would prevent us from adding useful suggestions like adding explicit lifetime bounds to the opaque type (I guess!). Maybe we can improve diagnostics by some other way, at least by removing the cryptic `id<T>::{opaque#0}`.

This is  problem for TAIT as well because it shows the same cryptic error message, but I think here we want to suggest the bound `OpaqueTy<T>: 'static` rather than `T: 'static`.

Reply:
Note that TAITs will be nameable, so, we do need to keep this path

Reply:
I'd leave it as it is for TAIT as well. I prefer suggesting `T: 'static` than `OpaqueTy<T>: 'static`. It's true that the later is more general but they're not really different unless `T` appears unconstrained in the hidden type, which should be a rare use case.

IMO, it's a cursed hack even for projection types. This is a diagnostic issue and should not be dealt with here. The diagnostics machinery should be smarter to suggest the two solutions.